dts/bindings/mtd/atmel,sam-flash.yaml

```diff
@@ -0,0 +1,109 @@
+# Copyright (c) 2023 Bjarki Arge Andreasen
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  This binding describes the Atmel SAM flash area layout.
+
+  The Atmel SAM flash area varies in write-block-size, memory area,
+  and the layout of erase-blocks.
+
+  E.g. the flash area layout of the ATSAM4E16C:
+
+    |--------------------|
+    | 8 Kbytes           |  erase block size = 2048
+    |--------------------|
+    | 8 Kbytes           |  erase block size = 2048
+    |--------------------|
+    | 48 Kbytes          |  erase block size = 4096
+    |--------------------|
+    | 64 Kbytes          |  erase block size = 4096
+    |--------------------|
+    | ...                |
+
+  The ATSAM4E16C has a flash area which is 1000Kbytes
+  (1024 * 1024 bytes) with a write-block-size of 8 bytes. The first
+  16 Kbytes can be erased in blocks of 2048 bytes
+  (8 blocks of 2048 bytes), the remaining flash area is erasable
+  in blocks of 4096 bytes (252 blocks of 4096 bytes).
+
+  This flash area layout would described as:
+
+    / {
+      soc {
+        eefc: flash-controller@400e0a00 {
+        compatible = "atmel,sam-flash-controller";
+        reg = <0x400e0a00 0x200>;
+        clocks = <&pmc PMC_TYPE_PERIPHERAL 6>;
+        status = "okay";
+
+        #address-cells = <1>;
+        #size-cells = <1>;
+        #erase-block-cells = <2>;
+
+        flash0: flash@400000 {
+          compatible = "atmel,sam-flash", "soc-nv-flash";
+          reg = <0x400000 0x100000>;
+          write-block-size = <8>;
+          erase-blocks = <&eefc 8 2048>, <&eefc 252 4096>;
+        };
+      };
+    };
+
+  Notes:
+    The flash area layout node flash0 should have both this
+    compatible, "atmel,sam-flash", and the "soc-nv-flash"
```

Comment:
mcuboot->MCUboot. And I am quite sure the MCUboot picks the flash device by `zephyr,code-partition`. The `soc-nv-flash` compatible may be used by some linker scripts, but that happens withing Zephyr.
Have you tried to build for example "Hello world", without MCUboot, and with that compatible to assigned to the node?

Reply:
I will try, my findings here are based on a samplesize of 1 :)

Reply:
OK, update on that. The "soc-nv-flash" is used by flash map macros to figure out which device actually is flash device.
If your driver makes sub-devices of controller real devices then lack of "soc-nv-flash" will work fine.
I have described the details here: https://github.com/zephyrproject-rtos/zephyr/pull/64089
In general when you try to do something to a flash device that hangs from flash-controller you have to do:
```
flash_<action>(DEVICE_DT_GET(DT_PARENT(DT_NODELABEL(flash0)))
```
rather than:
```
flash_<action>(DEVICE_DT_GET(DT_NODELABEL(flash0))
```
because flash-controller is a device and the flash0 thing hanging from it is just a configuration (?),
Flash Map needs such devices to be marked as "soc-nv-flash" to be able to get proper device pointer. We should probably have some other compatible string to address the issue.
The macro that is problematic is here: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/devicetree/fixed-partitions.h#L86